### PR TITLE
RFC - Make omniauth use post by default

### PR DESCRIPTION
**Pull request by BobbyMcWho**
Re:
https://github.com/omniauth/omniauth/issues/960
https://github.com/omniauth/omniauth/pull/809

Shoutout @pat, this work was influenced by https://github.com/omniauth/omniauth/pull/963

This would obviously need to be a major version bump due to breaking changes. 

This is based on the ideas of a few others, with some things I added in myself. 

The changes include:  
- Defaulting to POST only allowed request methods
- Use of a `request_validation_phase` specific to this type of workflow. 
- Default request validation phase that should work for _most_ rack based apps, though Rails is a special case 

Example apps available [here](https://github.com/BobbyMcWho/omniauth_2_examples). 

Rails apps will require a special [verification phase](https://github.com/BobbyMcWho/omniauth_2_examples/blob/main/rails_app.ru#L14-L28) that they [config](https://github.com/BobbyMcWho/omniauth_2_examples/blob/main/rails_app.ru#L34) in an initializer (or elsewhere if using devise). 

Seeking feedback and reviews, hoping I didn't miss anything glaringly obvious here.

### CODE REVIEW

**Comment by josh-m-sharpe** [comment]
Love the progress and direction.

Why not stick the rails specific configuration in a railtie so apps don't have to do anything?

**Comment by BobbyMcWho** [comment]
> Love the progress and direction!
>
> Why not stick the rails specific configuration in a railtie so apps don't have to do anything?

I'm not particularly fond of including Rails specific code in libraries, I know that is also a position held by past maintainers of this project as well. 

If rails changes the internals of the protect_from_request_forgery classes, that means that omniauth is now required to update and write version-specific implementations. 

I'd much rather leave it to the application owners to configure, or have the frequently used https://github.com/cookpad/omniauth-rails_csrf_protection updated post release of OmniAuth 2.0 to configure it.

**Review by danielpaul — DISMISSED** [review]

**Review by suprnova32 — APPROVED** [review]
Looking good :shipit:

**Review by suprnova32 — APPROVED** [review]

**Comment by BobbyMcWho** [comment]
Going to merge this into a 2_0_indev branch and see if there are any other breaking change chores that we wrap in while we're releasing a major version bump.

File: lib/omniauth/strategy.rb
Comment (BobbyMcWho) [reply lib/omniauth/strategy.rb:183]: Hmm, I see what you're saying, however it will only log in the case that `OmniAuth.config.allowed_request_methods.include?(:get)` resolves to true. (See line 198)

Comment (sponomarev) [reply lib/omniauth/strategy.rb:183]: Yeah, correct. During the upgrade process, we wanted to track down any potential cases where `GET` is used instead of the legit `POST`. We relied on the https://github.com/omniauth/omniauth/wiki/Upgrading-to-2.0#get that says

> Because using GET for login poses concerns (see OWASP Cheatsheet), after upgrading OmniAuth will log a :warn level log with **every GET request to a login path** when your OmniAuth.config.allowed_request_methods include :get.

Instead, we've uncovered that the warning is produced to any GET request to our app and the log stream is getting bloated pretty fast.

Should the `warn_if_using_get` be guarded with `on_auth_path?` check? We can work on the PR if you confirm.

Comment (BobbyMcWho) [reply lib/omniauth/strategy.rb:183]: I think that would be a reasonable resolution, yes. A PR would be appreciated and I'll hopefully have some time soon to go through the release w/ a patch bump for you.

Comment (charlie-wasp) [reply lib/omniauth/strategy.rb:183]: @BobbyMcWho Hi, I've prepared a PR for this issue https://github.com/omniauth/omniauth/pull/1041

**Comment by idrozd** [comment]
HI!

Something I was not able to find anywhere is any comments on whether enforcing POST for request phase implies that Single Sign On aka SSO would be affected:

As far as I understand, switching to POST only does make sense if the request phase form submission itself is explicit and imposed to user – it would not make sense to have some JS that submits the form automatically.

Imagine a situation where single auth provider is being used by ecosystem of apps – let's say each of these apps is a tab in a navigation bar of a portal. 
Switching between apps is currently seamless with GET request phase.

**Pull request merged**
